meilisearch-http/src/extractors/authentication/mod.rs

```diff
@@ -2,28 +2,85 @@ mod error;
 
 use std::marker::PhantomData;
 use std::ops::Deref;
+use std::pin::Pin;
 
 use actix_web::FromRequest;
 use futures::future::err;
-use futures::future::{ok, Ready};
-use meilisearch_error::ResponseError;
+use futures::Future;
+use meilisearch_error::{Code, ResponseError};
 
 use error::AuthenticationError;
 use meilisearch_auth::{AuthController, AuthFilter};
 
-pub struct GuardedData<T, D> {
+pub struct GuardedData<P, D> {
     data: D,
     filters: AuthFilter,
-    _marker: PhantomData<T>,
+    _marker: PhantomData<P>,
 }
 
-impl<T, D> GuardedData<T, D> {
+impl<P, D> GuardedData<P, D> {
     pub fn filters(&self) -> &AuthFilter {
         &self.filters
     }
+
+    async fn auth_bearer(
+        auth: AuthController,
+        token: String,
+        index: Option<String>,
+        data: Option<D>,
+    ) -> Result<Self, ResponseError>
+    where
+        P: Policy + 'static,
+    {
+        match Self::authenticate(auth, token, index).await? {
+            (_, Some(filters)) => match data {
+                Some(data) => Ok(Self {
+                    data,
+                    filters,
+                    _marker: PhantomData,
+                }),
+                None => Err(AuthenticationError::IrretrievableState.into()),
+            },
+            (token, None) => Err(AuthenticationError::InvalidToken(token).into()),
+        }
+    }
+
+    async fn auth_token(auth: AuthController, data: Option<D>) -> Result<Self, ResponseError>
+    where
+        P: Policy + 'static,
+    {
+        match Self::authenticate(auth, "", None).await?.1 {
+            Some(filters) => match data {
+                Some(data) => Ok(Self {
+                    data,
+                    filters,
+                    _marker: PhantomData,
+                }),
+                None => Err(AuthenticationError::IrretrievableState.into()),
+            },
+            None => Err(AuthenticationError::MissingAuthorizationHeader.into()),
+        }
+    }
+
+    async fn authenticate<S>(
+        auth: AuthController,
+        token: S,
+        index: Option<String>,
+    ) -> Result<(S, Option<AuthFilter>), ResponseError>
```

Comment:
We may return only an `Option<AuthFilter>`, this would simplify the upper functions.

Reply:
no sorry, I was saying:
```suggestion
    ) -> Result<Option<AuthFilter>, ResponseError>
```

Is it possible ?

Reply:
no, we need S to be static here, so it could force a clone. This is the only way to report errors without having to clone each time on line 44.

Reply:
On the lines 113, 114 and 115 you clone a lot, maybe you can simplify this method and also clone this token?

Reply:
The issue is not about cloning when it's needed, as it is on the lines you mention, but cloning when it is not, for error handling....

But ultimately, passing the token seems useless, I think it's better to just stop passing I it to the `InvalidToken`